pkg/strc/middleware_logger.go

```diff
@@ -0,0 +1,104 @@
+package strc
+
+import (
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	echoproxy "github.com/osbuild/logging/pkg/echo"
+)
+
+func slogAttributesFromRequest(r *http.Request) []slog.Attr {
+	attrs := []slog.Attr{
+		{
+			Key: "request",
+			Value: slog.GroupValue(
+				[]slog.Attr{
+					slog.String("method", r.Method),
+					slog.String("host", r.Host),
+					slog.String("path", r.URL.Path),
+					slog.String("user-agent", r.UserAgent()),
+					slog.String("ip", r.RemoteAddr),
+					slog.Int64("length", r.ContentLength),
+				}...,
+			),
+		},
+	}
+
+	traceID := TraceIDFromContext(r.Context())
+	if traceID != EmptyTraceID {
+		attrs = append(attrs, slog.String(TraceIDKey, traceID.String()))
+	}
+	spanID := SpanIDFromContext(r.Context())
+	if spanID != EmptySpanID {
+		attrs = append(attrs, slog.String(SpanIDKey, spanID.String()))
+	}
+
+	return attrs
+}
+
+func EchoRequestLogger(logger *slog.Logger, config MiddlewareConfig) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			for _, filter := range config.Filters {
+				if !filter(nil, c.Request()) {
+					return next(c)
+				}
+			}
+
+			c.SetLogger(echoproxy.NewProxyWithContextFor(logger, c.Request().Context()))
```

Comment:
Can you also extract this one? This should really do one and one thing only - request logging.

Reply:
sure thing